core/isolate.rs

```diff
@@ -742,8 +744,8 @@ impl Future for Isolate {
     inner.check_promise_errors();
     inner.check_last_exception()?;
 
-    // We're idle if pending_ops is empty.
-    if inner.pending_ops.is_empty() {
+    // We're idle if all pending_ops have blocks_exit flag false.
+    if inner.pending_ops.iter().all(|op| !op.1) {
```

Comment:
Not so sure if this would have some impact under large amount of concurrent async op (iterating potentially thousands of pending ops for each poll).

When I did my impl in #2735 , I used a simple counter: whenever optional op added, increment; when done, decrement; when number of pending ops is same as number of optional ops, we know all the remaining are optional ones and can safely exit.

Reply:
I added a todo comment for it.

I agree that this could cause perf issue in future, but I think it's not a huge problem for a moment because there is only a few optional async ops (0 at this moment, and 1 if #3610 landed).

Do you think that we need to address it before landing this?